src/kernl/implementations/attention.py

```diff
@@ -437,53 +496,33 @@ def forward(
             assert attention_mask.size(3) == size_n, "Last size of mask must broadcast on QK^t"
 
             HAS_MASK = True
-
-        _fwd_kernel[grid](
-            heads=heads,
-            size_m=size_m,
-            size_n=size_n,
-            size_m_rounded=size_m_rounded,
-            Q=q,
-            K=k,
-            V=v,
-            sm_scale=sm_scale,
-            attention_mask=attention_mask,
-            TMP=tmp,
-            output=output,
-            q_batch_stride=q.stride(0),
-            q_head_stride=q.stride(1),
-            q_m_stride=q.stride(2),
-            q_k_stride=q.stride(3),
-            k_batch_stride=k.stride(0),
-            k_head_stride=k.stride(1),
-            k_n_stride=k.stride(2),
-            k_k_stride=k.stride(3),
-            v_batch_stride=v.stride(0),
-            v_head_stride=v.stride(1),
-            v_k_stride=v.stride(2),
-            v_n_stride=v.stride(3),
-            o_batch_stride=output.stride(0),
-            o_head_stride=output.stride(1),
-            o_m_stride=output.stride(2),
-            o_n_stride=output.stride(3),
-            attention_mask_batch_stride=attention_mask.stride(0) if HAS_MASK else 0,
-            attention_mask_head_stride=attention_mask.stride(1) if HAS_MASK else 0,
-            attention_mask_m_stride=attention_mask.stride(2) if HAS_MASK else 0,
-            attention_mask_k_stride=attention_mask.stride(3) if HAS_MASK else 0,
-            NEED_LOAD_MASK_SIZE_N=NEED_LOAD_MASK_SIZE_N,
-            NEED_LOAD_MASK_SIZE_M=NEED_LOAD_MASK_SIZE_M,
-            min_clamp_value=torch.finfo(attention_mask.dtype).min if HAS_MASK else 0,
-            MASK_BATCH_SIZE=attention_mask.size(0) if HAS_MASK else 0,
-            MASK_HEAD_SIZE=attention_mask.size(1) if HAS_MASK else 0,
-            MASK_M_SIZE=attention_mask.size(2) if HAS_MASK else 0,
-            MASK_K_SIZE=attention_mask.size(3) if HAS_MASK else 0,
-            HAS_MASK=HAS_MASK,
-            IS_CAUSAL=is_causal,
-            BLOCK_M=BLOCK_M,
-            BLOCK_N=BLOCK_M,
-            BLOCK_DHEAD=dhead,
-            num_warps=4,
-            num_stages=1,
+            IS_MATRIX_MASK = attention_mask.size(2) != 1
+
+        _fwd_kernel[grid](  # can't use name args because of the way autotune is implemented :-(
+            heads,  # heads
+            size_m,  # size_m
+            size_n,  # size_n
+            size_m // 32,  # size_m_cache_key
+            size_n // 32,  # size_n_cache_key
+            q,  # Q
+            k,  # K
+            v,  # V
+            sm_scale,  # sm_scale
```

Comment:
why do you put useless comments when the name match parameter

Reply:
Homogeneous